要运行此代码，请在 **免费** 的 Tesla T4 Google Colab 实例上按 "*Runtime*" 并选择 "*Run all*"。

<div align="center">
<a href="https://unsloth.ai/"><img src="https://github.com/unslothai/unsloth/raw/main/images/unsloth%20new%20logo.png" width="115"></a>
<a href="https://discord.gg/unsloth"><img src="https://github.com/unslothai/unsloth/raw/main/images/Discord button.png" width="145"></a>
<a href="https://docs.unsloth.ai/"><img src="https://github.com/unslothai/unsloth/blob/main/images/documentation%20green%20button.png?raw=true" width="125"></a> 加入 Discord 获取帮助 + ⭐ <i>在 <a href="https://github.com/unslothai/unsloth">Github</a> 上给我们加星</i> ⭐
</div>

如果您想在自己的电脑上安装 Unsloth，请按照我们 [Github 页面](https://docs.unsloth.ai/get-started/installing-+-updating)上的安装说明操作。

您将学习如何进行 [数据准备](#Data)、如何 [训练](#Train)、如何 [运行模型](#Inference)、以及 [如何保存模型](#Save)。

### News

**阅读我们的 [Gemma 3 博客](https://unsloth.ai/blog/gemma3) 了解 Unsloth 的最新动态，以及我们的 [Reasoning 博客](https://unsloth.ai/blog/r1-reasoning) 了解如何训练推理模型。**

访问我们的文档，获取所有 [模型上传](https://docs.unsloth.ai/get-started/all-our-models) 和 [笔记本](https://docs.unsloth.ai/get-started/unsloth-notebooks)。

### Installation

In [1]:
%%capture
import os
if "COLAB_" not in "".join(os.environ.keys()):
    # 如果不是在 Colab 环境中运行，使用以下命令安装 Unsloth
    !pip install unsloth
else:
    # 仅在 Colab 笔记本中执行以下操作！否则请使用 pip install unsloth
    !pip install --no-deps bitsandbytes accelerate xformers==0.0.29.post3 peft trl triton cut_cross_entropy unsloth_zoo
    !pip install sentencepiece protobuf datasets huggingface_hub hf_transfer
    !pip install --no-deps unsloth


### Unsloth

In [2]:
from unsloth import FastLanguageModel
import torch

# 最大序列长度，可自由选择！我们内部自动支持 RoPE Scaling。
max_seq_length = 2048

# 数据类型设置，默认自动检测。Tesla T4 和 V100 使用 Float16，Ampere+ 使用 Bfloat16。
dtype = None

# 是否启用 4bit 量化以减少内存使用。可以设置为 False。
load_in_4bit = True

# 以下是我们支持的 4bit 预量化模型，下载速度提升 4 倍且避免内存不足问题。
fourbit_models = [
    "unsloth/Meta-Llama-3.1-8B-bnb-4bit",      # Llama-3.1 模型，15 万亿 tokens，速度提升 2 倍！
    "unsloth/Meta-Llama-3.1-8B-Instruct-bnb-4bit",
    "unsloth/Meta-Llama-3.1-70B-bnb-4bit",
    "unsloth/Meta-Llama-3.1-405B-bnb-4bit",    # 我们还上传了 405B 的 4bit 模型！
    "unsloth/Mistral-Nemo-Base-2407-bnb-4bit", # 新的 Mistral 12B，速度提升 2 倍！
    "unsloth/Mistral-Nemo-Instruct-2407-bnb-4bit",
    "unsloth/mistral-7b-v0.3-bnb-4bit",        # Mistral v3，速度提升 2 倍！
    "unsloth/mistral-7b-instruct-v0.3-bnb-4bit",
    "unsloth/Phi-3.5-mini-instruct",           # Phi-3.5，速度提升 2 倍！
    "unsloth/Phi-3-medium-4k-instruct",
    "unsloth/gemma-2-9b-bnb-4bit",
    "unsloth/gemma-2-27b-bnb-4bit",            # Gemma，速度提升 2 倍！
]
# 更多模型请访问 https://huggingface.co/unsloth

# 加载模型和分词器
model, tokenizer = FastLanguageModel.from_pretrained(
    # 可选择以下模型：
    # "unsloth/Qwen2.5-0.5B", "unsloth/Qwen2.5-1.5B", "unsloth/Qwen2.5-3B"
    # "unsloth/Qwen2.5-14B",  "unsloth/Qwen2.5-32B",  "unsloth/Qwen2.5-72B"
    # 以及所有 Instruct 版本和数学、编程版本！
    model_name = "unsloth/Qwen2.5-7B",
    max_seq_length = max_seq_length,
    dtype = dtype,
    load_in_4bit = load_in_4bit,
    # token = "hf_...", # 如果使用受限模型（如 meta-llama/Llama-2-7b-hf），需要提供 token。
)


🦥 Unsloth: Will patch your computer to enable 2x faster free finetuning.
🦥 Unsloth Zoo will now patch everything to make training faster!
==((====))==  Unsloth 2025.3.19: Fast Qwen2 patching. Transformers: 4.50.2.
   \\   /|    Tesla T4. Num GPUs = 1. Max memory: 14.741 GB. Platform: Linux.
O^O/ \_/ \    Torch: 2.6.0+cu124. CUDA: 7.5. CUDA Toolkit: 12.4. Triton: 3.2.0
\        /    Bfloat16 = FALSE. FA [Xformers = 0.0.29.post3. FA2 = False]
 "-____-"     Free license: http://github.com/unslothai/unsloth
Unsloth: Fast downloading is enabled - ignore downloading bars which are red colored!


model.safetensors.index.json:   0%|          | 0.00/106k [00:00<?, ?B/s]

model-00001-of-00002.safetensors:   0%|          | 0.00/5.00G [00:00<?, ?B/s]

model-00002-of-00002.safetensors:   0%|          | 0.00/2.54G [00:00<?, ?B/s]

Loading checkpoint shards:   0%|          | 0/2 [00:00<?, ?it/s]

generation_config.json:   0%|          | 0.00/172 [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/4.72k [00:00<?, ?B/s]

vocab.json:   0%|          | 0.00/2.78M [00:00<?, ?B/s]

merges.txt:   0%|          | 0.00/1.67M [00:00<?, ?B/s]

added_tokens.json:   0%|          | 0.00/605 [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/617 [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/11.4M [00:00<?, ?B/s]

我们现在添加了 LoRA 适配器，因此我们只需要更新全部参数的 1% 到 10%！

In [3]:
model = FastLanguageModel.get_peft_model(
    model,
    r = 16, # 可选择任意大于 0 的数值！推荐值：8、16、32、64、128
    target_modules = ["q_proj", "k_proj", "v_proj", "o_proj",
                      "gate_proj", "up_proj", "down_proj",],
    lora_alpha = 16,
    lora_dropout = 0, # 支持任意值，但设置为 0 是经过优化的
    bias = "none",    # 支持任意值，但设置为 "none" 是经过优化的
    # [新特性] "unsloth" 使用了 30% 更少的显存，可容纳 2 倍更大的批次！
    use_gradient_checkpointing = "unsloth", # 设置为 True 或 "unsloth" 以支持超长上下文
    random_state = 3407,
    use_rslora = False,  # 我们支持 Rank Stabilized LoRA
    loftq_config = None, # 以及 LoftQ
)


Unsloth 2025.3.19 patched 28 layers with 28 QKV layers, 28 O layers and 28 MLP layers.


<a name="Data"></a>
### 数据准备

我们现在使用来自 [yahma](https://huggingface.co/datasets/yahma/alpaca-cleaned) 的 Alpaca 数据集，这是原始 [Alpaca 数据集](https://crfm.stanford.edu/2023/03/13/alpaca.html) 的一个过滤版本，包含 52K 数据。你可以用自己的数据准备代码替换此部分。

**[注意]** 如果只想训练完成任务（忽略用户输入），请参考 TRL 的文档 [这里](https://huggingface.co/docs/trl/sft_trainer#train-on-completions-only)。

**[注意]** 请务必在分词后的输出中添加 **EOS_TOKEN**！否则会导致无限生成问题！

如果你想使用 `llama-3` 模板来处理 ShareGPT 数据集，可以尝试我们的对话式 [笔记本](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Llama3_(8B)-Alpaca.ipynb)。

对于文本生成任务（如小说写作），可以尝试这个 [笔记本](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Mistral_(7B)-Text_Completion.ipynb)。


In [9]:
alpaca_prompt = """你是一个信息安全专家，你精通车联网安全、物联网安全、Web安全、无线电安全、移动应用安全等领域，请详细认真的协助我完整各种工作，以下是一个描述任务的指令，与提供进一步上下文的输入配对。请写出一个适当完成请求的回应。

### 指令:
{}

### 输入:
{}

### 回应:
{}"""

EOS_TOKEN = tokenizer.eos_token # 必须添加 EOS_TOKEN
def formatting_prompts_func(examples):
    instructions = examples["instruction"]
    inputs       = examples["input"]
    outputs      = examples["output"]
    texts = []
    for instruction, input, output in zip(instructions, inputs, outputs):
        # 必须添加 EOS_TOKEN，否则生成过程将无限进行！
        text = alpaca_prompt.format(instruction, input, output) + EOS_TOKEN
        texts.append(text)
    return { "text" : texts, }
pass

from datasets import load_dataset
dataset = load_dataset("yahma/alpaca-cleaned", split = "train")
dataset = dataset.map(formatting_prompts_func, batched = True,)


NameError: name 'tokenizer' is not defined

<a name="Train"></a>
### 训练模型
现在让我们使用 Huggingface TRL 的 `SFTTrainer`！更多文档请参考：[TRL SFT 文档](https://huggingface.co/docs/trl/sft_trainer)。为了加快速度，我们仅进行 60 步训练，但你可以设置 `num_train_epochs=1` 来进行完整训练，并关闭 `max_steps=None`。我们也支持 TRL 的 `DPOTrainer`！


In [5]:
from trl import SFTTrainer
from transformers import TrainingArguments
from unsloth import is_bfloat16_supported

trainer = SFTTrainer(
    model = model,
    tokenizer = tokenizer,
    train_dataset = dataset,
    dataset_text_field = "text",
    max_seq_length = max_seq_length,
    dataset_num_proc = 2,
    packing = False, # 对于短序列，可以使训练速度提高 5 倍。
    args = TrainingArguments(
        per_device_train_batch_size = 2,
        gradient_accumulation_steps = 4,
        warmup_steps = 5,
        # num_train_epochs = 1, # 设置此项以进行一次完整的训练。
        max_steps = 60,
        learning_rate = 2e-4,
        fp16 = not is_bfloat16_supported(),
        bf16 = is_bfloat16_supported(),
        logging_steps = 1,
        optim = "adamw_8bit",
        weight_decay = 0.01,
        lr_scheduler_type = "linear",
        seed = 3407,
        output_dir = "outputs",
        report_to = "none", # 使用此项可集成到 WandB 等工具
    ),
)


Unsloth: Tokenizing ["text"] (num_proc=2):   0%|          | 0/51760 [00:00<?, ? examples/s]

In [8]:
# @title 显示当前内存统计信息
gpu_stats = torch.cuda.get_device_properties(0)
start_gpu_memory = round(torch.cuda.max_memory_reserved() / 1024 / 1024 / 1024, 3)
max_memory = round(gpu_stats.total_memory / 1024 / 1024 / 1024, 3)
print(f"GPU = {gpu_stats.name}。最大内存 = {max_memory} GB。")
print(f"{start_gpu_memory} GB 的内存已被保留。")


GPU = Tesla T4。最大内存 = 14.741 GB。
7.246 GB 的内存已被保留。


In [9]:
trainer_stats = trainer.train()

==((====))==  Unsloth - 2x faster free finetuning | Num GPUs used = 1
   \\   /|    Num examples = 51,760 | Num Epochs = 1 | Total steps = 60
O^O/ \_/ \    Batch size per device = 2 | Gradient accumulation steps = 4
\        /    Data Parallel GPUs = 1 | Total batch size (2 x 4 x 1) = 8
 "-____-"     Trainable parameters = 40,370,176/7,000,000,000 (0.58% trained)


Unsloth: Will smartly offload gradients to save VRAM!


In [10]:
# @title 显示最终内存和时间统计信息
used_memory = round(torch.cuda.max_memory_reserved() / 1024 / 1024 / 1024, 3)
used_memory_for_lora = round(used_memory - start_gpu_memory, 3)
used_percentage = round(used_memory / max_memory * 100, 3)
lora_percentage = round(used_memory_for_lora / max_memory * 100, 3)
print(f"训练耗时 {trainer_stats.metrics['train_runtime']} 秒。")
print(
    f"训练耗时 {round(trainer_stats.metrics['train_runtime']/60, 2)} 分钟。"
)
print(f"峰值保留内存 = {used_memory} GB。")
print(f"训练峰值保留内存 = {used_memory_for_lora} GB。")
print(f"峰值保留内存占最大内存的百分比 = {used_percentage} %。")
print(f"训练峰值保留内存占最大内存的百分比 = {lora_percentage} %。")


训练耗时 451.0191 秒。
训练耗时 7.52 分钟。
峰值保留内存 = 9.381 GB。
训练峰值保留内存 = 2.135 GB。
峰值保留内存占最大内存的百分比 = 63.639 %。
训练峰值保留内存占最大内存的百分比 = 14.483 %。


<a name="Inference"></a>
### 推理
让我们运行模型吧！你可以修改指令和输入——将输出留空即可！

**[新增] 在免费的 Colab 中体验 Llama-3.1 8b Instruct 的2倍速推理 [点击这里](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Unsloth_Studio.ipynb)**


In [11]:
# alpaca_prompt = 从上方复制
FastLanguageModel.for_inference(model)  # 启用原生2倍速推理
inputs = tokenizer(
[
    alpaca_prompt.format(
        "继续斐波那契数列。",  # 指令
        "1, 1, 2, 3, 5, 8",  # 输入
        "",  # 输出 - 留空以生成结果！
    )
], return_tensors = "pt").to("cuda")

outputs = model.generate(**inputs, max_new_tokens = 64, use_cache = True)
tokenizer.batch_decode(outputs)


['以下是一个描述任务的指令，与提供进一步上下文的输入配对。请写出一个适当完成请求的回应。\n\n### 指令:\n继续斐波那契数列。\n\n### 输入:\n1, 1, 2, 3, 5, 8\n\n### 回应:\n13, 21, 34, 55, 89, 144, 233, 377, 610, 987, 1597, 2584, 4181, 6']

*斜体文本* 你还可以使用 `TextStreamer` 进行连续推理——这样你可以逐个查看生成的 token，而无需等待整个生成过程完成！

In [12]:
# alpaca_prompt = 从上方复制
FastLanguageModel.for_inference(model)  # 启用原生2倍速推理
inputs = tokenizer(
[
    alpaca_prompt.format(
        "继续斐波那契数列。",  # 指令
        "1, 1, 2, 3, 5, 8",  # 输入
        "",  # 输出 - 留空以生成结果！
    )
], return_tensors = "pt").to("cuda")

from transformers import TextStreamer
text_streamer = TextStreamer(tokenizer)
_ = model.generate(**inputs, streamer=text_streamer, max_new_tokens=128)


以下是一个描述任务的指令，与提供进一步上下文的输入配对。请写出一个适当完成请求的回应。

### 指令:
继续斐波那契数列。

### 输入:
1, 1, 2, 3, 5, 8

### 回应:
13, 21, 34, 55, 89, 144, 233, 377, 610, 987, 1597, 2584, 4181, 6765, 10946, 17711, 28657, 46368, 75025, 121393, 196418, 317811, 


<a name="Save"></a>
### 保存和加载微调模型
要将最终模型保存为 LoRA 适配器，可以使用 Huggingface 的 `push_to_hub` 进行在线保存，或者使用 `save_pretrained` 进行本地保存。

**[注意]** 这仅保存 LoRA 适配器，而不是完整模型。如果需要保存为 16bit 或 GGUF 格式，请向下滚动查看！


In [13]:
model.save_pretrained("lora_model")  # 本地保存
tokenizer.save_pretrained("lora_model")
# model.push_to_hub("your_name/lora_model", token = "...")  # 在线保存
# tokenizer.push_to_hub("your_name/lora_model", token = "...")  # 在线保存


('lora_model/tokenizer_config.json',
 'lora_model/special_tokens_map.json',
 'lora_model/vocab.json',
 'lora_model/merges.txt',
 'lora_model/added_tokens.json',
 'lora_model/tokenizer.json')

现在，如果你想加载我们刚刚保存的 LoRA 适配器进行推理，将 `False` 设置为 `True`：

In [14]:
if False:
    from unsloth import FastLanguageModel
    model, tokenizer = FastLanguageModel.from_pretrained(
        model_name = "lora_model",  # 你用于训练的模型名称
        max_seq_length = max_seq_length,
        dtype = dtype,
        load_in_4bit = load_in_4bit,
    )
    FastLanguageModel.for_inference(model)  # 启用原生2倍速推理

# alpaca_prompt = 你必须从上方复制！

inputs = tokenizer(
[
    alpaca_prompt.format(
        "巴黎著名的高塔是什么？",  # 指令
        "",  # 输入
        "",  # 输出 - 留空以生成结果！
    )
], return_tensors = "pt").to("cuda")

from transformers import TextStreamer
text_streamer = TextStreamer(tokenizer)
_ = model.generate(**inputs, streamer=text_streamer, max_new_tokens=128)


以下是一个描述任务的指令，与提供进一步上下文的输入配对。请写出一个适当完成请求的回应。

### 指令:
巴黎著名的高塔是什么？

### 输入:


### 回应:
巴黎著名的高塔是埃菲尔铁塔。它位于法国巴黎的战神广场，是世界上最著名的建筑之一，也是法国的象征。埃菲尔铁塔高324米，由古斯塔夫·埃菲尔设计，于1889年建成，是当时世界上最高的建筑物。现在，它仍然是世界上最受欢迎的旅游景点之一，每年吸引数百万游客前来参观。<|endoftext|>


你也可以使用 Hugging Face 的 `AutoModelForPeftCausalLM`。仅当你没有安装 `unsloth` 时使用它。不过，这种方式可能会非常慢，因为不支持下载 `4bit` 模型，并且 Unsloth 的**推理速度快 2 倍**。

In [ ]:
if False:
    # 我强烈不建议使用此方法 - 如果可能，请使用 Unsloth
    from peft import AutoPeftModelForCausalLM
    from transformers import AutoTokenizer
    model = AutoPeftModelForCausalLM.from_pretrained(
        "lora_model",  # 你用于训练的模型名称
        load_in_4bit = load_in_4bit,
    )
    tokenizer = AutoTokenizer.from_pretrained("lora_model")


### 保存为 float16 以用于 VLLM

我们还支持直接保存为 `float16` 格式。选择 `merged_16bit` 保存为 float16，或者选择 `merged_4bit` 保存为 int4。我们还允许使用 `lora` 适配器作为备用方案。使用 `push_to_hub_merged` 可以将模型上传到你的 Hugging Face 账户！你可以访问 https://huggingface.co/settings/tokens 获取你的个人令牌。

In [ ]:
# 合并为 16bit
if False: model.save_pretrained_merged("model", tokenizer, save_method = "merged_16bit",)
if False: model.push_to_hub_merged("hf/model", tokenizer, save_method = "merged_16bit", token = "")

# 合并为 4bit
if False: model.save_pretrained_merged("model", tokenizer, save_method = "merged_4bit",)
if False: model.push_to_hub_merged("hf/model", tokenizer, save_method = "merged_4bit", token = "")

# 仅保存 LoRA 适配器
if False: model.save_pretrained_merged("model", tokenizer, save_method = "lora",)
if False: model.push_to_hub_merged("hf/model", tokenizer, save_method = "lora", token = "")


### GGUF / llama.cpp 转换

现在我们原生支持保存为 `GGUF` / `llama.cpp` 格式！我们会克隆 `llama.cpp` 并默认保存为 `q8_0`。同时支持所有量化方法，例如 `q4_k_m`。使用 `save_pretrained_gguf` 可以进行本地保存，使用 `push_to_hub_gguf` 可以上传到 Hugging Face。

一些支持的量化方法（完整列表请参考我们的[Wiki 页面](https://github.com/unslothai/unsloth/wiki#gguf-quantization-options)）：
* `q8_0` - 快速转换。资源使用较高，但通常可以接受。
* `q4_k_m` - 推荐。对于 attention.wv 和 feed_forward.w2 张量的一半使用 Q6_K，其余部分使用 Q4_K。
* `q5_k_m` - 推荐。对于 attention.wv 和 feed_forward.w2 张量的一半使用 Q6_K，其余部分使用 Q5_K。

[**新功能**] 如果需要微调并自动导出到 Ollama，可以试试我们的 [Ollama notebook](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Llama3_(8B)-Ollama.ipynb)。

In [ ]:
# 保存为 8bit Q8_0
if False: model.save_pretrained_gguf("model", tokenizer,)
# 请记得访问 https://huggingface.co/settings/tokens 获取一个 token！
# 并将 hf 替换为你的用户名！
if False: model.push_to_hub_gguf("hf/model", tokenizer, token = "")

# 保存为 16bit GGUF
if False: model.save_pretrained_gguf("model", tokenizer, quantization_method = "f16")
if False: model.push_to_hub_gguf("hf/model", tokenizer, quantization_method = "f16", token = "")

# 保存为 q4_k_m GGUF
if False: model.save_pretrained_gguf("model", tokenizer, quantization_method = "q4_k_m")
if False: model.push_to_hub_gguf("hf/model", tokenizer, quantization_method = "q4_k_m", token = "")

# 保存为多个 GGUF 配置选项 - 如果需要多个格式，速度更快！
if False:
    model.push_to_hub_gguf(
        "hf/model", # 将 hf 替换为你的用户名！
        tokenizer,
        quantization_method = ["q4_k_m", "q8_0", "q5_k_m",],
        token = "", # 在 https://huggingface.co/settings/tokens 获取 token
    )


现在，你可以在 `llama.cpp` 或基于用户界面的系统（例如 Jan 或 Open WebUI）中使用 `model-unsloth.gguf` 文件或 `model-unsloth-Q4_K_M.gguf` 文件。你可以在以下链接安装 Jan：[点击这里](https://github.com/janhq/jan)，或者安装 Open WebUI：[点击这里](https://github.com/open-webui/open-webui)。

到此就完成了！如果你对 Unsloth 有任何疑问，我们有一个 [Discord](https://discord.gg/unsloth) 频道！如果你发现任何问题、希望获取最新的大语言模型（LLM）动态、需要帮助、参与项目等，欢迎加入我们的 Discord！

其他相关链接：
1. 训练你自己的推理模型 - Llama GRPO 笔记本 [免费 Colab](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Llama3.1_(8B)-GRPO.ipynb)
2. 保存微调模型到 Ollama - [免费笔记本](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Llama3_(8B)-Ollama.ipynb)
3. Llama 3.2 Vision 微调 - 放射学应用场景 [免费 Colab](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Llama3.2_(11B)-Vision.ipynb)
4. 查看关于 DPO、ORPO、持续预训练、对话微调等的更多笔记本，请访问我们的[文档](https://docs.unsloth.ai/get-started/unsloth-notebooks)。

<div class="align-center">
  <a href="https://unsloth.ai"><img src="https://github.com/unslothai/unsloth/raw/main/images/unsloth%20new%20logo.png" width="115"></a>
  <a href="https://discord.gg/unsloth"><img src="https://github.com/unslothai/unsloth/raw/main/images/Discord.png" width="145"></a>
  <a href="https://docs.unsloth.ai/"><img src="https://github.com/unslothai/unsloth/blob/main/images/documentation%20green%20button.png?raw=true" width="125"></a>

  如果需要帮助，请加入 Discord + ⭐️ <i>在 <a href="https://github.com/unslothai/unsloth">Github</a> 上为我们点亮星星</i> ⭐️
</div>